Continue absorbing cardano-api in Direct.Tx, baby steps -- albeit big baby.

KtorZ opened the pull request:
- :round_pushpin: **Add smart-constructors and utilities to Hydra.Ledger.Cardano**
  
- :round_pushpin: **Define extra helpers for building transactions in steps.**
  Co-authored-by: ch1bo <sebastian.nagel@ncoding.at>

- :round_pushpin: **Make use of the new transaction builder API for 'initTx'**
  Co-authored-by: ch1bo <sebastian.nagel@ncoding.at>

- :round_pushpin: **Make use of the transaction builder API for 'commitTx'**
    Keeping the same strategy as for initTx and leaving the signature untouched. Note that this currently fails because 'unsafeBuildTransaction' throws with 'TxBodyEmptyTxInsCollateral'; looks like cardano-api does this extra verification and requires collateral to be set when creating the transaction... which is unfortunate as we do set the collateral only during transaction balancing.

- :round_pushpin: **Work around guards of 'makeTransactionBody' to keep the ball rolling :grimacing:**
    The proper way to fix this (and we'll get there eventually) will be to get the tiny wallet work with a 'TxBuilder' instead of a plain 'CardanoTx', so that 'unsafeBuildTransaction' is only called when the transaction is actually ready to be built. Right now, we do however return full transaction (albeit not passing phase-1 validations...) in order to test their execution. So, having 'initTx', 'commitTx', 'abortTx'... return a 'TxBuilder' is probably what we want (then bypassing the guards only in the relevant test code) but it's a bit too much of a step right now.

ch1bo committed:
Add smart-constructors and utilities to Hydra.Ledger.Cardano

KtorZ committed:
Define extra helpers for building transactions in steps.

Co-authored-by: ch1bo <sebastian.nagel@ncoding.at>

KtorZ committed:
Make use of the new transaction builder API for 'initTx'

Co-authored-by: ch1bo <sebastian.nagel@ncoding.at>

KtorZ committed:
Make use of the transaction builder API for 'commitTx'

  Keeping the same strategy as for initTx and leaving the signature untouched. Note that this currently fails because 'unsafeBuildTransaction' throws with 'TxBodyEmptyTxInsCollateral'; looks like cardano-api does this extra verification and requires collateral to be set when creating the transaction... which is unfortunate as we do set the collateral only during transaction balancing.

KtorZ committed:
Work around guards of 'makeTransactionBody' to keep the ball rolling :grimacing:

  The proper way to fix this (and we'll get there eventually) will be to get the tiny wallet work with a 'TxBuilder' instead of a plain 'CardanoTx', so that 'unsafeBuildTransaction' is only called when the transaction is actually ready to be built. Right now, we do however return full transaction (albeit not passing phase-1 validations...) in order to test their execution. So, having 'initTx', 'commitTx', 'abortTx'... return a 'TxBuilder' is probably what we want (then bypassing the guards only in the relevant test code) but it's a bit too much of a step right now.

KtorZ commented:
@abailly-iohk I'd be tempted to have an `AntiCorruption` module also factoring out all the `fromLedger / toLedger` stuff, and have it re-exposed by the Ledger.Cardano one. What do you think?

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Tx.hs:
```diff
@@ -176,56 +178,54 @@ withOutputs newOutputs txbody =
+
+  mkInitial (Api.toPlutusKeyHash . Api.verificationKeyHash -> vkh) =
+    Api.TxOut initialAddress initialValue initialDatum
+   where
```
Should: Remove this `where`. It is not required and the definitions below could live just fine in the same `where` as `mkInitial`. I'd argue having less indentation / scopes is better to read/modify.

KtorZ replied:
I would disagree! Nesting `where` is one of the best feature of the language IMO as it helps keeping things tidy and showing that some bits of code are only related to some other bits and not others. As soon as a function starts having more than 5 `where` or so, it becomes hard to keep track of what is relevant when. Grouping `where` clauses as such is helping readability and brain load a lot!

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Tx.hs:
```diff
@@ -176,56 +178,54 @@ withOutputs newOutputs txbody =
-  initialDatum vkey = Data . toData $ MockInitial.datum $ pubKeyHash vkey
+  headOut =
+    Api.TxOut headAddress headValue headDatum
+   where
```
Should: Also remove this `where` (see other comment for rationale)

ch1bo commented on hydra-node/src/Hydra/Ledger/Cardano.hs:
```diff
@@ -221,6 +224,89 @@ utxoFromTx (Tx body@(ShelleyTxBody _ ledgerBody _ _ _ _) _) =
+-- with scripts but no collaterals. This is unfortunate because collaterals are
+-- currently added after by out integrated wallet... We may want to revisit our
+-- flow to avoid this exception and have the wallet work from a TxBuilder instead
+-- of fiddling with a sealed 'CardanoTx'.
```
This is what we discussed once and not the fault of `emptyTxBody` I think :)

ch1bo commented on hydra-node/src/Hydra/Ledger/Cardano.hs:
```diff
@@ -221,6 +224,89 @@ utxoFromTx (Tx body@(ShelleyTxBody _ ledgerBody _ _ _ _) _) =
+-- Transaction Construction
+--
+
+type TxBuilder = TxBodyContent BuildTx Era
```
I like the name, but introducing this, essentially has readers / editors jump through one more indirection when they would want to find out what that type actually is, e.g. what record fields it has.

KtorZ replied:
I was originally thinking creating an actual newtype out of it, and make this newtype a `Monoid` because the current interface kind of screams for it... but for the sake of doing baby steps and keeping things simple, I didn't. The alias stayed.

ch1bo commented on hydra-node/src/Hydra/Ledger/Cardano.hs:
```diff
@@ -401,22 +485,65 @@ fromLedgerTxId :: Ledger.TxId Ledger.StandardCrypto -> TxId
+-- discriminant which is currently fully captured as 'Mainnet | Testnet'.
+--
+-- So, it would be a slightly better DX to use Mainnet | Testnet as an interface
+-- here since we are only constructing Shelley addresses.
```
Good input :+1:

ch1bo commented on hydra-node/src/Hydra/Ledger/Cardano.hs:
```diff
@@ -474,9 +624,41 @@ fromCardanoApiUtxo :: Cardano.Api.UTxO AlonzoEra -> Utxo
+-- Witnesses
 --
 
+-- TODO: This could be made available upstream...
```
But we don't need it now, right? We could just always create a `Witness ScriptWitnessForSpending Era` for now, right? i.e. we could generalize "our helper" `mkScriptWitness` when we need to and then, this might already be merged upstream?

KtorZ replied:
Well, since we already know we are going to need script witnesses for minting, I went for the generalized version right away (which isn't much more work to be honest...).

ch1bo commented on hydra-node/src/Hydra/Ledger/Cardano.hs:
```diff
@@ -497,6 +679,18 @@ fromLedgerTxWitness wits =
+
+liftValue :: Value -> TxOutValue Era
+liftValue =
+  TxOutValue MultiAssetInAlonzoEra
```
Should: find a better name. This one is confusing IMO. Analogous to `mkTxOutDatum` this could be named `mkTxOutValue`

ch1bo replied:
Which raises the question that maybe we want to have a `mkTxOut :: Plutus.ToData a => ... -> Value -> a -> TxOut CtxTx Era` ?

ch1bo commented on hydra-node/src/Hydra/Ledger/Cardano.hs:
```diff
@@ -497,6 +679,18 @@ fromLedgerTxWitness wits =
+
+fromLedgerValue :: Ledger.Mary.Value Ledger.StandardCrypto -> Value
+fromLedgerValue =
+  fromMaryValue
```
Why not rename `fromMaryValue` to `fromLedgerValue`?

KtorZ replied:
`fromMaryValue` comes from the `cardano-api` ...

ch1bo reviewed (changes requested):
Only minor naming things, which we could discuss quickly in the ensemble.

KtorZ committed:
Avoid too nested where clauses.

  As discussed, we want to avoid nested where but, we also don't want to
  clutter the module scope and haddock doc for our modules.
